rts/motoko-rts/src/gc/mark_compact.rs

```diff
@@ -0,0 +1,312 @@
+        }
+
+        TAG_VARIANT => {
+            let variant = obj as *mut Variant;
+            push_mark_stack((*variant).field, heap_base);
+        }
+
+        TAG_CONCAT => {
+            let concat = obj as *mut Concat;
+            push_mark_stack(concat.text1(), heap_base);
+            push_mark_stack(concat.text2(), heap_base);
+        }
+
+        TAG_OBJ_IND => {
+            let obj_ind = obj as *mut ObjInd;
+            push_mark_stack((*obj_ind).field, heap_base);
+        }
+
+        TAG_BITS64 | TAG_BITS32 | TAG_BLOB | TAG_BIGINT => {
+            // These don't include pointers, skip
+        }
+
+        TAG_NULL => {
+            rts_trap_with("encountered NULL object tag in dynamic object in mark_fields");
+        }
+
+        TAG_FWD_PTR | _ => {
+            // Any other tag is a bug
+            rts_trap_with("invalid object tag in mark_fields");
+        }
+    }
+}
+
+unsafe fn thread_roots(static_roots: SkewedPtr, heap_base: u32) {
+    // Static roots
+    let root_array = static_roots.as_array();
+    for i in 0..root_array.len() {
+        thread_obj_fields(root_array.get(i).unskew() as *mut Obj, heap_base);
+    }
+    // No need to thread closure table here as it's on heap and we already marked it
+}
+
+/// Scan the heap, update forward references. At the end of this pass all fields will be threaded
+/// and forward references will be updated, pointing to the object's new location.
+unsafe fn update_fwd_refs(heap_base: u32) {
+    let mut free = heap_base;
+
+    for bit in iter_bits() {
+        let p = (heap_base + (bit * WORD_SIZE)) as *mut Obj;
+
+        // Update forward references to the object to the object's new location and restore
+        // object header
+        unthread(p, free);
+
+        // Thread fields
+        thread_obj_fields(p, heap_base);
+
+        free += object_size(p as usize).to_bytes().0;
+    }
+}
+
+/// Expects all fields to be threaded. Updates backward references and moves objects to their new
+/// locations.
+unsafe fn update_bwd_refs(heap_base: u32) {
+    let mut free = heap_base;
+
+    for bit in iter_bits() {
+        let p = (heap_base + (bit * WORD_SIZE)) as *mut Obj;
+
+        // Update backward references to the object's new location and restore object header
+        unthread(p, free);
+
+        // All references to the object now point to the new location, move the object
+        let p_size_words = object_size(p as usize);
+        if free as usize != p as usize {
+            memcpy_words(free as usize, p as usize, p_size_words);
+        }
+
+        free += p_size_words.to_bytes().0;
+    }
+
+    crate::gc::HP = free;
+}
+
+unsafe fn thread_obj_fields(obj: *mut Obj, heap_base: u32) {
+    match obj.tag() {
+        TAG_OBJECT => {
+            let obj = obj as *mut Object;
+            let obj_payload = obj.payload_addr();
+            for i in 0..obj.size() {
+                thread(obj_payload.add(i as usize), heap_base);
+            }
+        }
```

Comment:
There is some repetition here and in other places.

How about a higher-order function (sorry for the non-Rusty syntax)
```
with_ptrs_in_obj : (*mut Obj) -> (*mut SkewedPtr -> ()) -> ()
```
that you can use is in `thread_obj_fields` and `mark_fields` and in the copying GC? (Assuming Rust will inline it, of course)? I think this would greatly improve readability, and also avoid bugs as we change the object layout or add new object types.

Reply:
IIRC I had tried this before but it did not work well because of some difficulty with handing `BigInt`s. Now that we have a simpler `BigInt` representation I think it should work out nicely. I will update.

Reply:
I implemented this in 9af5279 . It improved performance (instructions used) slightly in copying GC (< 1%) but made compacting GC 3% worse. The code is much better though..